diff -Naur /old/boost/numeric/ublas/experimental/sparse_view.hpp /new/boost/numeric/ublas/experimental/sparse_view.hpp
--- /old/boost/numeric/ublas/experimental/sparse_view.hpp	2012-09-06 10:10:56.000000000 +0200
+++ /new/boost/numeric/ublas/experimental/sparse_view.hpp	2012-09-06 10:12:05.000000000 +0200
@@ -190,11 +190,11 @@
 
         BOOST_UBLAS_INLINE
         compressed_matrix_view(const compressed_matrix_view& o) :
-            size1_(size1_), size2_(size2_),
-            nnz_(nnz_),
-            index1_data_(index1_data_),
-            index2_data_(index2_data_),
-            value_data_(value_data_)
+            size1_(o.size1_), size2_(o.size2_),
+            nnz_(o.nnz_),
+            index1_data_(o.index1_data_),
+            index2_data_(o.index2_data_),
+            value_data_(o.value_data_)
         {}
 
         //
